src/sentry/static/sentry/app/api.jsx

```diff
@@ -235,18 +235,24 @@ export class Client {
   }
 
   requestPromise(path, {includeAllArgs, ...options} = {}) {
-    return new Promise((resolve, reject) => {
-      // Create an error object here before we make any async calls so
-      // that we have a helpful stacktrace if it errors
-      const error = new Error('API Request Failed');
+    // Create an error object here before we make any async calls so
+    // that we have a helpful stacktrace if it errors
+    const error = new Error(`${options.method || 'GET'} "${path}"`);
 
+    return new Promise((resolve, reject) => {
       this.request(path, {
         ...options,
         success: (data, ...args) => {
           includeAllArgs ? resolve([data, ...args]) : resolve(data);
         },
         error: (resp, ...args) => {
+          // Update error message with response status code
+          error.message = `${error.message} -> ${(resp && resp.status) || 'n/a'}`;
           error.resp = resp;
+
+          // Drop the Client.requestPromise frame so stacktrace starts at `requestPromise` callsite
+          const lines = error.stack.split('\n');
+          error.stack = [lines[0], ...lines.slice(2)].join('\n');
```

Comment:
How bad is this @HazAT @kamilogorek ? Would `in_app_include` accomplish something similar?

Reply:
Yeah it probably would, I will write down that we should implement this option in the JS SDK.
Given all other options, this approach seems to be a reasonable way to go.

Reply:
Yeah it probably would, I will write down that we should implement this option in the JS SDK.
Given all other options, this approach seems to be a reasonable way to go.

Reply:
I'd approach this like this:

```js
class RequestError extends Error {
  constructor(method, path, resp) {
    const message = `${method || 'GET'} "${path}" -> ${resp && resp.status || 'n/a'}`;
    super(message);

    this.message = message;
    this.resp = resp;

    Error.captureStackTrace(this, this.constructor)
    const stack = this.stack.split('\n');
    this.stack = [stack[0], ...stack.slice(2)].join('\n');

    this.name = new.target.prototype.constructor.name;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

// and then

return new Promise((resolve, reject) => {
  this.request(path, {
    error: (resp, ...args) => reject(new RequestError(options.method, path, resp))
  })
})
```

Oooor

```js
class RequestError extends Error {
  constructor(method, path, resp) {
    const message = `${method || 'GET'} "${path}" -> ${resp && resp.status || 'n/a'}`;
    super(message);

    this.message = message;
    this.resp = resp;

    Error.captureStackTrace(this, this.constructor)
    const stack = this.stack.split('\n');
    this.stack = [stack[0], ...stack.slice(2)].join('\n');

    this.name = new.target.prototype.constructor.name;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

// and then

Sentry.init({
  beforeSend(event, hint) {
    if (hint && hint.originalException instanceof RequestError) {
      // Remove 2nd frame to cleanup unnecessary Promise call
      try { exception.values[0].stacktrace.frames.splice(1, 1); } catch (oO) {}
    }
    return event;
  }
});

// ^ or make it an eventProcessor instead
```

Reply:
Damn @kamilogorek

Reply:
Thinking about it a bit more, this is highly coupled to the `requestPromise`, I wouldn't want it in the SDK's eventProcessor. I like the RequestError approach although we can't create the error in  the error callback of `request` because it's in a new async stack.

Reply:
Ah, we have a babel plugin that allows extending builtin `Error` -- I'm just gonna keep it as-is to reduce possibilities of the stack changing.